stan/math/prim/fun/lb_constrain.hpp

```diff
@@ -20,22 +23,17 @@ namespace math {
  *
  * <p>where \f$L\f$ is the constant lower bound.
  *
- * <p>If the lower bound is negative infinity, this function
- * reduces to <code>identity_constrain(x)</code>.
- *
- * @tparam T type of scalar
+ * @tparam T type of Matrix
  * @tparam L type of lower bound
- * @param[in] x Unconstrained scalar input
- * @param[in] lb lower bound on constrained output
- * @return lower bound constrained value corresponding to inputs
+ * @param[in] x Unconstrained Matrix input
+ * @param[in] lb Lower bound
+ * @return Constrained matrix
  */
 template <typename T, typename L>
-inline return_type_t<T, L> lb_constrain(const T& x, const L& lb) {
-  using std::exp;
-  if (lb == NEGATIVE_INFTY) {
-    return identity_constrain(x);
-  }
-  return exp(x) + lb;
+inline auto lb_constrain(const T& x, const L& lb) {
+  const auto& lb_ref = to_ref(lb);
+  check_finite("lb_constrain", "lb", value_of(lb_ref));
+  return eval(add(exp(x), lb));
```

Comment:
We don't need the eval here now

Reply:
Honestly not sure. `to_ref` might allocate memory, so if this were to be assigned to a non-heavy type, then it would need a holder to work. I will ditch the evals though to see if tests pass. @t4c1 can you advise on this?

Reply:
Follow up [here](https://github.com/stan-dev/math/pull/2285#discussion_r553677134)

Reply:
`lb_ref ` is a local variable referenced in returned expression, so it needs a holder.

```
return make_holder([&x](auto& lb_ref){
  check_finite("lb_constrain", "lb", value_of(lb_ref));
  return exp(x) + lb_ref;
}, to_ref(std::forward<L>(lb)));
```

Also change `const L& lb` to `L&& lb`.

Reply:
I get segfaults with this:

```cpp
template <typename T, typename L, typename U>
inline promote_scalar_type_t<return_type_t<T, L, U>, T>
lub_constrain(const T& x, const L& lb, const U& ub,
              return_type_t<T, L, U>& lp) {
  const auto& x_ref = to_ref(x);
  const auto& lb_ref = to_ref(lb);
  const auto& ub_ref = to_ref(ub);

  check_less("lub_constrain", "lb", value_of(lb_ref), value_of(ub_ref));
  check_finite("lub_constrain", "lb", value_of(lb_ref));
  check_finite("lub_constrain", "ub", value_of(ub_ref));

  const auto& diff = to_ref(subtract(ub_ref, lb_ref));

  lp += sum(add(log(diff),
                subtract(-abs(x_ref), multiply(2, log1p_exp(-abs(x_ref))))));

  return make_holder([](const auto& diff, const auto& x_ref, const auto& lb_ref) {
      return add(elt_multiply(diff, inv_logit(x_ref)), lb_ref);
    }, diff, x_ref, lb_ref);
}
```

It seems to be associated with the temporary `const auto& diff = to_ref(subtract(ub_ref, lb_ref));`.

I'm just going to code these to use plain types. I don't understand the assumptions around holders and don't feel safe using them myself and I don't think where these functions are used expressions matter.

Reply:
Alright I ended up going with evals to get the plain types. I think it's correct. eval means, if there are expressions, make them plain types.

`promote_scalar_t` doesn't work right with varmat (you get matvars from varmats). `return_var_matrix_t` doesn't work right with non-reverse mode stuff.

Reply:
That segfaults because you are passing lvalues into `make_holder`, so it thinks they won't go out of scope and does not construct a holder. Instead you should `std::move` these variables into `make_holder`.

Reply:
This was the issue. They have been successfully holder-ified.